### ISSUE TITLE
Using SMACK for Rust code

### ISSUE THREAD

**Issue by cmr** [smackers/smack#219]
Sort of a meta-issue I guess.

Depends on #165, probably #54.

I'm experimenting with this now.

**Comment by zvonimir** [comment]
We have an ongoing preliminary effort that is extending SMACK to be able to use it on Rust. If you are interested in this direction and would like to contribute, please let us know. There is certainly a lot of work still to be done :).

**Comment by ZackPierce** [comment]
I'm curious about the current state of the preliminary efforts to support Rust.  Has anything been done beyond the https://github.com/smackers/smack/compare/rust-support prototype branch from a year ago?

I've tossed up a super minimal rust project within the [examples directory on a branch](https://github.com/ZackPierce/smack/tree/rusty), accompanied by a helper script that mimics some core bits of the SMACK toolchain. Currently the process dies in the middle of llvm2bpl.

**Comment by HiDefender** [comment]
I'm interested in making this happen and have the time to devote to it. Where is the best place to start?

**Comment by zvonimir** [comment]
@HiDefender, @ZackPierce, @cmr : FYI, we finally have a preliminary working (kind of) support for Rust in the rust-prelims branch. Feel free to try it out. I am sure that we'll identify a lot of outstanding issues and missing features in the next several weeks, and help with these would be greatly appreciated.

**Comment by HiDefender** [comment]
@zvonimir Research has gone a different direction. Pretty sure I'll still need something like SMACK for Rust in the future, but for the time being I need to get my preliminaries working.

**Comment by ruffsl** [comment]
Hello @zvonimir , I'd like jump in and try using smack with rust for information flow control. I see there are some ifc examples in the `rust-prelims` branch, with the latest commit aa871e5 on Jul 19, 2018.

I also see there is the branches:
* `modular-rust` with the latest commit f1630a0 on Jan 24, 2018
* `rust-llvm-7` with the latest commit ba405c6 on Jul 31, 2019.

For those just getting to know smack, rust, an IFC, where would be a good point to start, and what is the state/progress/purpose of each rust related branch thus far?

**Comment by zvonimir** [comment]
Hey @ruffsl and thanks for your interest in SMACK. So currently the `rust-llvm-7` is the most up to date branch with support for Rust. Note that it actually supports LLVM 8 :).

We are actively working on Rust support, and I expect for this branch to be merged into develop soon. But that branch would be a good place to start. We should also create a simple tutorial for how to run SMACK on Rust, so keep your eyes open for that as well.

Finally, we would love to collaborate on this effort, and any feedback you might have for us would be greatly appreciated!

**Comment by zvonimir** [comment]
So our develop and master branches now have prototypical support for Rust, and anyone interested in it can give it a try. This is still work in progress, but it is probably time we close this issue.
